controllers/utils/multicloud_util.go

```diff
@@ -0,0 +1,360 @@
+		if errors.IsNotFound(err) {
+			multiCloudLog.Info("[EnsureManagedClusterActionForInstallPlan] MCA doesn't exist, create it",
+				"name", name, "namespace", namespace)
+			actionMeta := metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			}
+			actionSpec, err := NewManagedClusterActionForInstallPlanSpec(installPlan)
+			if err != nil {
+				return nil, err
+			}
+			mcaForInstallPlan = &actionv1beta1.ManagedClusterAction{
+				ObjectMeta: actionMeta,
+				Spec:       *actionSpec,
+			}
+
+			if err := c.Create(ctx, mcaForInstallPlan); err != nil {
+				return mcaForInstallPlan, nil
+			}
+		} else {
+			return nil, err
+		}
+	} else {
+		// If the specific managedClusterView was found, update it.
+		multiCloudLog.Info("[EnsureManagedClusterActionForInstallPlan] MCA already exists, update it",
+			"name", name, "namespace", namespace)
+		actionSpec, err := NewManagedClusterActionForInstallPlanSpec(installPlan)
+		if err != nil {
+			return nil, err
+		}
+		mcaForInstallPlan.Spec = *actionSpec
+
+		if err := c.Update(ctx, mcaForInstallPlan); err != nil {
+			return nil, err
+		}
+	}
+
+	return mcaForInstallPlan, nil
+}
+
+// NewManagedClusterActionForInstallPlanSpec returns the action spec for approving an InstallPlan.
+func NewManagedClusterActionForInstallPlanSpec(installPlan operatorsv1alpha1.InstallPlan) (*actionv1beta1.ActionSpec, error) {
+	installPlanClusterServiceVersionNames, err := json.Marshal(installPlan.Spec.ClusterServiceVersionNames)
+	if err != nil {
+		return nil, fmt.Errorf("Error trying to Marshal InstallPlan ClusterServiceVersionNames: %s", err.Error())
+	}
+
+	templateContent := fmt.Sprintf(
+		`{"apiVersion": "operators.coreos.com/v1alpha1","kind": "InstallPlan",
+		  "metadata": {"name": "%s","resourceVersion": "%s"},
+		  "spec": {"approval": "Manual","approved": true, "clusterServiceVersionNames": %s}}`,
+		installPlan.ObjectMeta.Name,
+		installPlan.ObjectMeta.ResourceVersion,
+		string(installPlanClusterServiceVersionNames))
+
+	actionSpec := actionv1beta1.ActionSpec{
+		ActionType: actionv1beta1.UpdateActionType,
+		KubeWork: &actionv1beta1.KubeWorkSpec{
+			Resource:       "installplan",
+			Namespace:      installPlan.ObjectMeta.Namespace,
+			ObjectTemplate: runtime.RawExtension{Raw: []byte(templateContent)},
+		},
+	}
+
+	return &actionSpec, nil
+}
+
+// CheckManagedClusterActionDone returns error for all the cases in which an action didn't manage to complete
+// successfully.
+func CheckManagedClusterActionDone(
+	ctx context.Context, c client.Client, mcaName string, mcaNamespace string, timeout int) error {
+	// Check the managedClusterAction has completed successfully.
+	mca := &actionv1beta1.ManagedClusterAction{}
+
+	// Do incremental checks until we can decide if the ManagedClusterAction was successful or not.
+	elapsedTimeSec := 0
+	waitTimeSec := ManagedClusterActionWaitTimeSec
+	runCount := 1
+
+	for elapsedTimeSec <= timeout {
```

Comment:
In a heavily loaded system the amount of time it takes to make API calls be be several seconds. This can really skew the logic of waiting based on adding the sleep time. In this case maybe up to ~25 additional seconds which may be OK but we should look to improve in a follow-on.